logstash-core/src/main/java/org/logstash/plugins/factory/PluginFactoryExt.java

```diff
@@ -0,0 +1,264 @@
+        return plugin(
+                RubyUtil.RUBY.getCurrentContext(), PluginLookup.PluginType.CODEC,
+                name.asJavaString(), source, (Map<String, IRubyObject>) args, pluginArgs
+        );
+    }
+
+    @Override
+    public Codec buildDefaultCodec(String codecName) {
+        return (Codec) JavaUtil.unwrapJavaValue(plugin(
+                RubyUtil.RUBY.getCurrentContext(), PluginLookup.PluginType.CODEC,
+                codecName, null, Collections.emptyMap(), Collections.emptyMap()
+        ));
+    }
+
+    @SuppressWarnings("unchecked")
+    @JRubyMethod(required = 3, optional = 1)
+    public IRubyObject plugin(final ThreadContext context, final IRubyObject[] args) {
+        return plugin(
+                context,
+                PluginLookup.PluginType.valueOf(args[0].asJavaString().toUpperCase(Locale.ENGLISH)),
+                args[1].asJavaString(),
+                JavaUtil.unwrapIfJavaObject(args[2]),
+                args.length > 3 ? (Map<String, IRubyObject>) args[3] : new HashMap<>(),
+                null
+        );
+    }
+
+    @SuppressWarnings("unchecked")
+    private IRubyObject plugin(final ThreadContext context, final PluginLookup.PluginType type, final String name,
+                               SourceWithMetadata source, final Map<String, IRubyObject> args,
+                               Map<String, Object> pluginArgs) {
+        final String id = generateOrRetrievePluginId(context, type, name, source);
+        pluginsById.add(id);
+        final AbstractNamespacedMetricExt typeScopedMetric = metrics.create(context, type.rubyLabel());
+
+        final PluginLookup.PluginClass pluginClass = pluginResolver.resolve(type, name);
+        if (pluginClass.language() == PluginLookup.PluginLanguage.RUBY) {
+
+            final Map<String, Object> newArgs = new HashMap<>(args);
+            newArgs.put("id", id);
+            final RubyClass klass = (RubyClass) pluginClass.klass();
+            final ExecutionContextExt executionCntx = executionContext.create(
+                    context, RubyUtil.RUBY.newString(id), klass.callMethod(context, "config_name")
+            );
+            final RubyHash rubyArgs = RubyHash.newHash(context.runtime);
+            rubyArgs.putAll(newArgs);
+            if (type == PluginLookup.PluginType.OUTPUT) {
+                return new OutputDelegatorExt(context.runtime, RubyUtil.RUBY_OUTPUT_DELEGATOR_CLASS).initialize(
+                        context,
+                        new IRubyObject[]{
+                                klass, typeScopedMetric, executionCntx,
+                                OutputStrategyExt.OutputStrategyRegistryExt.instance(context, null),
+                                rubyArgs
+                        }
+                );
+            } else if (type == PluginLookup.PluginType.FILTER) {
+                return filterDelegator(
+                        context, null,
+                        new IRubyObject[]{
```

Comment:
:nail_care: would just change these to receive `IRubyObject...` as last arg (not much important)